### PR TITLE
subvolume: create and take advantage of copy-on-write

### PR DESCRIPTION
**Pull request by iam-TJ**
Ensure a subvolume is created.

When the output subvol and workspace/root are in the same BTRFS take advantage of copy-on-write and snapshots by making root a subvol and taking a snapshot of it for the output before deleting workspace.

Closes: https://github.com/systemd/mkosi/issues/1525
Closes: https://github.com/systemd/mkosi/issues/1534

### CODE REVIEW

File: mkosi.md
Comment (behrmann) [inline mkosi.md:334]: What's the reasoning behind adding `subvolume_ro`? Couldn't this be exchanged for just doing `subvolume` and have people who need that set the read-only flag in a finalise script? I'm asking because the extra tracking of two different output formats seems like added complexity that might be good to avoid.

**Review by DaanDeMeyer — CHANGES_REQUESTED** [review]
This is not the way to go, we should just always try to snapshot and if it fails fall back to a regular copy.

In fact, I don't think we should have a subvolume output format. We should just automatically create them if we're on btrfs. I would modify copy_path() to try to snapshot the source if the to the destination if the destination does not exist yet. If the destination already exists, I would try to btrfs send receive the source into the destination. If that fails because the source is not a subvolume, then we do a regular copy.

**Comment by iam-TJ** [comment]
Currently `copy-path()` isn't called at all when using `subvolume` which is part of the problem with recent changes having broken subvolume completely as reported in #1525

Extract from a build log using:
```
$ git l -n 4
38e6134 2023-05-03 11:54:48 +0100 N Tj fix: prefix must be valid machine name
ca8a943 2023-05-02 10:22:48 +0200 N Tj distro: debian: allow RemovePackages
3008c7e 2023-05-01 13:16:54 +0200 E Daan De Meyer Merge pull request #1513 from keszybz/more-man
be290f2 2023-05-01 13:15:33 +0200 E Daan De Meyer Merge pull request #1506 from DaanDeMeyer/profiles
```
shows no call to `copy_path()` when building a subvolume:
```
...
$ grep copy_path /tmp/mkosi/mkosi.36.log 
‣ copy_path('/home/tj/iam/mkosi/debian-amd64-base/mkosi.skeleton', '/home/tj/iam/mkosi/debian-amd64-base/mkosi.workspace/.mkosi.tmpzzhk_1bd/root, 'False', 'False')
‣ copy_path('/home/tj/iam/mkosi/debian-amd64-base/mkosi.extra', '/home/tj/iam/mkosi/debian-amd64-base/mkosi.workspace/.mkosi.tmpzzhk_1bd/root, 'False', 'False')
‣ copy_path('/home/tj/iam/mkosi/debian-amd64-base/mkosi.nspawn', '/home/tj/iam/mkosi/debian-amd64-base/mkosi.workspace/.mkosi.tmpzzhk_1bd/staging/debian-bullseye-amd64-0.2.nspawn, 'False', 'True')
```

**Comment by iam-TJ** [comment]
On 04/05/2023 09:24, Daan De Meyer wrote:
> Yes, so we shouldn't have a subvolume output format, we should have a --subvolume option, which makes us use subvolumes for certain directories:

That's a major change so I'll leave that to you.

I've been trying to re-instate broken functionality in order to continue 
working with existing project definitions.

**Comment by DaanDeMeyer** [comment]
Let's close this, I have an implementation of --subvolume locally already, will publish it after #1538 is merged.

**Pull request closed without merging**
